curvefs/src/client/fuse_s3_client.cpp

```diff
@@ -244,52 +244,48 @@ CURVEFS_ERROR FuseS3Client::Truncate(Inode *inode, uint64_t length) {
     return s3Adaptor_->Truncate(inode, length);
 }
 
-CURVEFS_ERROR FuseS3Client::FuseOpFlush(fuse_req_t req, fuse_ino_t ino,
-                                        struct fuse_file_info *fi) {
-    LOG(INFO) << "FuseOpFlush, ino: " << ino;
-
-    if (curvefs::client::common::FLAGS_enableCto) {
-        // need retry until success
-        while (CURVEFS_ERROR::OK != FuseOpFsync(req, ino, 0, fi)) {
-            sleep(option_.flushRetryIntervalMS / 1000);
-        }
-
-        return CURVEFS_ERROR::OK;
-    } else {
-        return FuseOpFsync(req, ino, 0, fi);
-    }
-}
-
 CURVEFS_ERROR FuseS3Client::FuseOpRelease(fuse_req_t req, fuse_ino_t ino,
-                                        struct fuse_file_info *fi) {
+                                          struct fuse_file_info *fi) {
     LOG(INFO) << "FuseOpRelease, ino: " << ino;
     CURVEFS_ERROR ret = CURVEFS_ERROR::OK;
 
+    // if cto, flush all write cache and clear read cache
     if (::curvefs::client::common::FLAGS_enableCto) {
-        ret = FuseOpFsync(req, ino, 0, fi);
+        ret = s3Adaptor_->FlushAllCache(ino);
         if (ret != CURVEFS_ERROR::OK) {
-            LOG(INFO) << "FuseOpRelease, ino: " << ino
-                      << " do fsync error: " << ret;
+            LOG(ERROR) << "FuseOpRelease, flush all cache fail, ret = " << ret
+                       << ", inodeid = " << ino;
             return ret;
         }
     }
 
     std::shared_ptr<InodeWrapper> inodeWrapper;
     ret = inodeManager_->GetInode(ino, inodeWrapper);
     if (ret != CURVEFS_ERROR::OK) {
-        LOG(ERROR) << "inodeManager get inode fail, ret = " << ret
-                   << ", inodeid = " << ino;
+        LOG(ERROR) << "FuseOpRelease, inodeManager get inode fail, ret = "
+                   << ret << ", inodeid = " << ino;
         return ret;
     }
 
     ::curve::common::UniqueLock lgGuard = inodeWrapper->GetUniqueLock();
+    // if cto, need sync inode meta
+    if (::curvefs::client::common::FLAGS_enableCto) {
+        ret = inodeWrapper->Sync();
+        if (ret != CURVEFS_ERROR::OK) {
+            LOG(ERROR) << "FuseOpRelease, inode sync s3 chunk info fail, ret = "
+                       << ret << ", inodeid = " << ino;
+            return ret;
+        }
+    }
+
     ret = inodeWrapper->Release();
     if (ret != CURVEFS_ERROR::OK) {
-        LOG(ERROR) << "inodeManager release inode fail, ret = " << ret
-                   << ", inodeid = " << ino;
+        LOG(ERROR) << "FuseOpRelease, inodeManager release inode fail, ret = "
+                   << ret << ", inodeid = " << ino;
         return ret;
     }
 
+    // if cto, need clear inode cache
     if (::curvefs::client::common::FLAGS_enableCto) {
         inodeManager_->ClearInodeCache(ino);
```

Comment:
if someone else open this inode concurrently and write，then clear inode here，will it cause a problem？